cpp/core/compute/ExecutionCtx.h

```diff
@@ -47,11 +48,18 @@ class ExecutionCtx : public std::enable_shared_from_this<ExecutionCtx> {
   ExecutionCtx(const std::unordered_map<std::string, std::string>& confMap) : confMap_(confMap) {}
   virtual ~ExecutionCtx() = default;
 
-  virtual std::shared_ptr<ResultIterator> getResultIterator(
+  virtual ResourceHandle createResultIterator(
       MemoryManager* memoryManager,
       const std::string& spillDir,
       const std::vector<std::shared_ptr<ResultIterator>>& inputs,
       const std::unordered_map<std::string, std::string>& sessionConf) = 0;
+  virtual ResourceHandle addResultIterator(std::shared_ptr<ResultIterator>) = 0;
+  virtual std::shared_ptr<ResultIterator> getResultIterator(ResourceHandle) = 0;
+  virtual void releaseResultIterator(ResourceHandle) = 0;
+
+  virtual ResourceHandle addBatch(std::shared_ptr<ColumnarBatch>) = 0;
+  virtual std::shared_ptr<ColumnarBatch> getBatch(ResourceHandle) = 0;
+  virtual void releaseBatch(ResourceHandle) = 0;
```

Comment:
I think there was a way to use WeakReference / PhantomReference / `finalize` API / `Cleaner` API to implement Java-to-C++ object binding. We can see if the way can be leveraged by us in future.